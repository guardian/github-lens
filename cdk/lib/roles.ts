import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuRole } from '@guardian/cdk/lib/constructs/iam';
import { CfnOutput, CfnParameter, Tags } from 'aws-cdk-lib';
import type { App } from 'aws-cdk-lib';
import {
	AccountPrincipal,
	Effect,
	Policy,
	PolicyStatement,
} from 'aws-cdk-lib/aws-iam';

export class CdkMetadataAccess extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const cdkMetadataAccountId = '095768028460';

		const stack = new CfnParameter(this, 'Stack', {
			description: 'Riffraff stack',
		});

		Tags.of(this).add('Stack', stack.valueAsString); // Needed to override stack info at runtime.

		const role = new GuRole(this, 'CdkMetadataRole', {
			roleName: 'cdk-metadata-access',
			description: 'Role CdkMetadata uses to crawl resources in this account',
			assumedBy: new AccountPrincipal(cdkMetadataAccountId),
		});

		new Policy(this, 'CdkMetadataPolicy', {
			policyName: 'CdkMetadataCollection',
			roles: [role],
			statements: [
				new PolicyStatement({
					effect: Effect.ALLOW,
					resources: ['*'],
					actions: [
						'cloudformation:Describe*',
						'cloudformation:Get*',
						'cloudformation:List*',
					],
				}),
			],
		});

		new CfnOutput(this, 'Role', {
			value: role.roleArn,
			description: 'CdkMetadata Role',
		});
	}
}
