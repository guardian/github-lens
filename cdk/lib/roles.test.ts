import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CdkMetadataAccess } from './roles';

describe('The Roles stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new CdkMetadataAccess(app, 'roles', {
			stack: 'deploy',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
