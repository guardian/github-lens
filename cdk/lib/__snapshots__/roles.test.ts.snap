// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Roles stack matches the snapshot 1`] = `
Object {
  "Outputs": Object {
    "Role": Object {
      "Description": "CdkMetadata Role",
      "Value": Object {
        "Fn::GetAtt": Array [
          "CdkMetadataRole1FF70BE2",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "Stack": Object {
      "Description": "Riffraff stack",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CdkMetadataPolicyD9889917": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "cloudformation:Describe*",
                "cloudformation:Get*",
                "cloudformation:List*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CdkMetadataCollection",
        "Roles": Array [
          Object {
            "Ref": "CdkMetadataRole1FF70BE2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CdkMetadataRole1FF70BE2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "arn:aws:iam::095768028460:root",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role CdkMetadata uses to crawl resources in this account",
        "RoleName": "cdk-metadata-access",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk-metadata",
          },
          Object {
            "Key": "Stack",
            "Value": Object {
              "Ref": "Stack",
            },
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
